sockeye/beam_search.py

```diff
@@ -83,10 +85,17 @@ def decode_step(self,
                     vocab_slice_ids: Optional[pt.Tensor] = None,
                     target_prefix_factor_mask: Optional[pt.Tensor] = None,
                     factor_vocab_size: Optional[int] = None):
-        logits, states, target_factor_outputs = self._model.decode_step(step_input, states, vocab_slice_ids)
+        logits, knn_probs, states, target_factor_outputs = self._model.decode_step(step_input, states, vocab_slice_ids)
         if not self._skip_softmax:
-            logits = pt.log_softmax(logits, dim=-1)
-        scores = -logits  # shape: (batch*beam, output_vocab_size/len(vocab_slice_ids))
+            if knn_probs is None:  # no knn used
+                probs = pt.log_softmax(logits, dim=-1)
+            else:
+                probs = pt.log(self.knn_lambda * pt.softmax(logits, dim=-1) + (1-self.knn_lambda) * knn_probs)
```

Comment:
have you considered moving this to a traced function? Might be worth benchmarking at some point.

Reply:
You could then also share the interpolation code between `SingleModelInference` and `EnsembleModelInference`

Reply:
re: tracing -- Yes. Initially we put them in a traced function and added `jit.ignore` to the function that did kNN query (because faiss operations should not be traced). Turns out in that case kNN query is only done once during the initial computational graph compilation and then the same `knn_probs` is reused for all subsequent segments, which is obviously wrong. So the kNN query and the interpolation were later moved out of the traced function (see [this commit](https://github.com/awslabs/sockeye/pull/1062/commits/b9592d44cf638925052adf20a9b79f4e3472f997)).

I cannot find concrete quote from the document or any Googled result but my guess is that `jit.ignore` only works with TorchScript but not tracing.

Reply:
Have you tried scripting the layer instead of tracing it? If you run `jit.script(layer)` when a layer is created, it should remain scripted (instead of traced) when its parent module is traced.

Reply:
I didn't -- will try!

Reply:
Hi Felix and Michael -- it's been a while since I last worked on this! I looked into the possibility of scripting this module -- unfortunately even this is not possible because `faiss.Index` is not a supported Torchscript type so it cannot be passed as an argument into the constructor. Here is the part of the error message that explains this:

```
Module 'KNN' has no attribute 'keys_index' (This attribute exists on the Python module, but we failed to convert Python type: 'faiss.swigfaiss.IndexFlat' to a TorchScript type. Only tensors and (possibly nested) tuples of tensors, lists, or dictsare supported as inputs or outputs of traced functions, but instead got value of type IndexFlat..
```

Hence, there is no way we can pass a faiss index into a scripted module, as all attributes of a scripted module have to be set in the constructor (otherwise they won't be scripted). So it seems like we'll have to keep this part of code as-is.

I've already addressed the other comments and will update the PR soon.